src/Std/Sat/CNF/Literal.lean

```diff
@@ -0,0 +1,47 @@
+/-
+Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Josh Clune
+-/
+prelude
+import Init.Data.ToString
+import Std.Sat.Basic
+
+namespace Std
+namespace Sat
+
+/--
+CNF literals identified by some type `α`.
+-/
+abbrev Literal (α : Type u) := α × Bool
+
+namespace Literal
+
+instance [Hashable α] : Hashable (Literal α) where
+  hash := fun x => if x.2 then hash x.1 else hash x.1 + 1
+
+instance : HSat α (Literal α) where
+  eval := fun p l => (p l.1) = l.2
+
+instance (p : α → Bool) (l : Literal α) : Decidable (p ⊨ l) := by
+  rw [HSat.eval, instHSat]
+  exact Bool.decEq (p l.fst) l.snd
```

Comment:
Constructing `Decidable` with tactics is a bad idea (remember #4674), so this should be
```suggestion
instance (p : α → Bool) (l : Literal α) : Decidable (p ⊨ l) :=
  inferInstanceAs (Decidable (p l.1 = l.2))
```
or use `decidable_of_decidable_of_iff`.